### PR TITLE
[SPARK-27385][SQL]add mapPartitionWithIndex to Dataframe

### PR DESCRIPTION
**Pull request by Ivankings**
modify comment

## What changes were proposed in this pull request?

This pr add mapPartitionsWithIndex to Dataframe.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by Ivankings** [comment]
If I want to use mapPartitionsWithIndex in Dataframe df,  I have to use df.rdd.mapPartitionsWithIndex instade. What I want to do is to use mapPartitionsWithIndex directly in Dataframe. So, I add mapPartitionsWithIndex in Dataframe, that's all.

**Comment by HeartSaVioR** [comment]
I'm not sure this is going to be approved, but I got same needs and had to handle with RDD, so for me this is good to add.

@Ivankings You should point to master branch, which major work happens. Spark 1.6 is already EOL.

**Comment by Ivankings** [comment]
@HeartSaVioR Thanks for that.

**Comment by kiszk** [comment]
@Ivankings Why do you make a PR to branch-1.6? Do you want to implement this feature as only for backport?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -1427,6 +1427,15 @@ class DataFrame private[sql](
     rdd.mapPartitions(f)
   }
 
+  /**
+    * Returns a new RDD by applying a function to each partition of this DataFrame, while tracking the index
+    * of the original partition.
+    * @group rdd
+    */
+  def mapPartitionsWithIndex[R: ClassTag](f: Iterator[Row] => Iterator[R]): RDD[R] = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Except for branch-1.6 issue others pointed, doesn't `rdd.mapPartitionsWithIndex` accepts `f: (Int, Iterator[T]) => Iterator[U]`? The give function doesn't have the partition index parameter?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Yeah I expected partition index being exposed to DataFrame API, not just calling rdd.mapPartitionsWithIndex which finally ends up with RDD. It's just a shortcut and high chance to be rejected.

**Pull request closed without merging**